### PR TITLE
volumewatcher: set maximum batch size for raft update

### PR DESCRIPTION
**Pull request by tgross**
For https://github.com/hashicorp/nomad/issues/7838

The `volumewatcher` has a 250ms batch window so claim updates will not typically be large enough to risk exceeding the maximum raft message size. But large jobs might have enough volume claims that this could be a danger. Set a maximum batch size of 100 messages per batch (roughly 31K), as a very conservative safety/robustness guard.

In the future I'd like to factor this same logic out of the `deploymentwatcher` batcher so we could share implementations.

### CODE REVIEW

File: nomad/volumewatcher/batcher.go [hunk 1]
```diff
@@ -67,30 +86,62 @@ func (b *VolumeUpdateBatcher) batcher() {
...
 				req.Claims = append(req.Claims, claim)
 			}
 
 			// Upsert the claims in a go routine
 			go f.Set(b.raft.UpsertVolumeClaims(&req))
 
-			// Reset the claims list and timer
-			claims = make(map[string]structs.CSIVolumeClaimRequest)
+			// Reset the batches list and timer
+			batches = batches[1:]
```
Comment (cgbaker) [inline nomad/volumewatcher/batcher.go:144]: if i'm reading this correctly, it looks like the loop relies on work coming in on `b.workCh` to reset the timer, same as before. however, with batching, it seems like it's possible that enough work could come in during a single `batchDuration` to create multiple batches. if that happens, the `timerCh` will signal a claim request for only the first batch of claims. the rest will not be dispatched until more work comes in to reset the timer (which could hypothetically be never).

so either this loop needs to send off all batches or it needs to reset `timerCh` (to `batchDuration` or perhaps some appropriate fraction thereof) if `len(batches) > 0`.

Comment (cgbaker) [reply nomad/volumewatcher/batcher.go:144]: aside: while i appreciate the elegance of this approach, why just not use a `Ticker` for `timerCh`?

Comment (tgross) [reply nomad/volumewatcher/batcher.go:144]: > so either this loop needs to send off all batches or it needs to reset timerCh (to batchDuration or perhaps some appropriate fraction thereof) if len(batches) > 0.

Good catch, that wasn't a problem when we were sending off the whole batch at once!

> aside: while i appreciate the elegance of this approach, why just not use a Ticker for timerCh?

Hm, that's a good point. In the original design (copied right out of [`deploymentwatcher`](https://github.com/hashicorp/nomad/blob/master/nomad/deploymentwatcher/batcher.go)), a no-op pass would potentially modify some state by swapping out the future. A ticker would make it a little easier to understand at the negligible cost of ticking over and checking the length each pass.

Comment (tgross) [reply nomad/volumewatcher/batcher.go]: I've updated with these changes. How's that look?

File: nomad/volumewatcher/batcher.go [hunk 2]
```diff
@@ -54,44 +61,83 @@ type updateWrapper struct {
...
-	future := NewBatchFuture()
+
+	// we track claimBatches rather than a slice of
+	// CSIVolumeClaimBatchRequest so that we can deduplicate updates
+	// for the same volume
+	batches := []*claimBatch{{
+		claims: make(map[string]structs.CSIVolumeClaimRequest),
+		future: NewBatchFuture(),
+	}}
+	ticker := time.NewTicker(b.batchDuration)
```
Comment (cgbaker) [inline nomad/volumewatcher/batcher.go:79]:
```suggestion
	ticker := time.NewTicker(b.batchDuration)
	defer ticker.Stop()
```

Comment (cgbaker) [reply nomad/volumewatcher/batcher.go:79]: https://golang.org/pkg/time/#NewTicker
> Stop the ticker to release associated resources.

**Pull request merged**